### DISCUSSION TITLE
tuples???

### DISCUSSION THREAD

**Discussion by PyAntony** [jasondelaat/pymonad#7] [General]
Hi Jason,

Thank you for all your work on this library. I am reading all the implementations and your code is so clear I am impressed. Your code is really a tutorial to understand monads. I have read about and used monads for long already, but never actually come across this concept of "all monads can be represented as tuples". Can you point me to other articles/videos to explore more about this perspective? Thanks!

**Comment by jasondelaat** [comment]
Thanks for the kind words!

I think I first came across the idea in a video by Bartosz Milewski which was part of a course he taught on categoy theory for programmers. Playlist is [here](https://www.youtube.com/playlist?list=PLbgaMIhjbmEnaH_LTkxLI7FMa2HsnawM_).

I don't remember exactly which video it was but [this one](https://www.youtube.com/watch?v=i9CU4CuHADQ&list=PLbgaMIhjbmEnaH_LTkxLI7FMa2HsnawM_&index=6) is my best guess. And he's probably also talked about it on his [website](https://bartoszmilewski.com/2014/10/28/category-theory-for-programmers-the-preface/) somewhere though, again, I'm not entirely sure which actual page.

I'm not entirely sure he ever actually explained it to be honest. I think he just sort of tossed the fact out and then moved on. But basically it has to do with algebraic data types. (He has a [video on that too.](https://www.youtube.com/watch?v=w1WMykh7AxA&list=PLbgaMIhjbmEnaH_LTkxLI7FMa2HsnawM_&index=10))

I don't know how much you know about ADTs so I apologize if I'm explaining things that you already understand but I'll try to give a brief explanation as I understand it. Essentially ADTs allow us to take a data type and transform it into an algebraic expression: something like ax^2 + b, where 'a', 'x', '2', and 'b' are all types. a, x, and b are all just general types that could be anything but '2' is how we represent booleans, the reason being booleans have exactly two values: True and False. In fact, '2' represents _any_ type with exactly two values because they're all isomorphic to each other. So: On/Off, Black/White, Stop/Go, etc. are all represented by the type '2'.

- Product types (basically tuples) are represented by multiplication in the expression.
- Sum types (types with multiple constructors) are represented by addition.
- Function types are represented by exponentiation.

That last one requires a bit of explanation. I won't explain why, just how to read it. In the above expression there's an x^2. That represents a function which takes a boolean as input and returns something of type 'x' (whatever type 'x' is.) So the exponent is the function's input type and the base is its output type.

For simplicity let's say that a = Int, b = Float, and x = String. Then the expression ax^2 + b is equivalent to the (Haskell) data type:

~~~haskell 
data Alg = Alg1 (Int, Bool -> String)
         | Alg2 Float 
~~~

I'm using Haskell types just because they're more compact.

Anyway, the point is: Once we've transformed a type into an algebraic expression we can manipulate it using all the normal rules of algebra and then transform it back into a different but equivalent type. And that's basically what's been done with most monad types, they've been transformed into a more convenient to use or more semantically meaningful form. But they're still equivalent to some tuple somewhere.

A couple examples:

Maybe:
1. The underlying tuple type is (x, Bool) where 'x' can be any type. True in the second position means the value is valid and False means it's invalid.
2. As an algebraic expression this is equivalent to 2x (two times x).
3. By normal algebraic rules we know that 2x = x + x
4. We can translate that back into a type:

~~~haskell 
data AlmostMaybe x = Just x -- corresponds to (x, True)
                   | Nothing x -- corresponds to (x, False)
~~~

Notice that since the type is x + x both constructors expect an 'x' as input. This is where semantics come into play because the semantics of Nothing are that whatever value we end up with is garbage and should be ignored. So we don't lose anything by just dropping the 'x' and getting the actual definition of Maybe:

~~~haskell 
data Maybe x = Just x
             | Nothing
~~~

State:
1. Underlying tuple is two functions. The current state to a value and the current state to a new state: (s -> a, s -> s)
2. As an algebraic expression: (a^s)(s^s) (a^s times s^s)
3. Since we're multiplying them and they have the same exponent we know that: (a^s)(s^s) = (as)^s
4. Which as a type is a function s -> (a, s), the current state to a tuple of the value and the new state, which is the normal representation of the state monad.
   
And so on. You can play this game with every monad.

Hopefully that helps and makes some kind of sense.

**Reply by PyAntony** [reply]
wow! I have never seen Monads from this perspective. This is by far the best description I have ever had. It would be really easy if we just think about Monads as tuples + the machinery to cleverly combine them (map/bind/then). Thank you for this clear explanation!

**Reply by jasondelaat** [reply]
Glad to help! It's worth mentioning that I actually messed up the implementation of a couple of the monads in this respect. Reader, I think, is one example. The implementations as they are work but they're not all necessarily 'correct' in this sense. So if you're trying to figure out how they all map to tuples, don't rely on my code to tell you the right way!

I'll fix them eventually but since they work for now it's not a high priority.
